### PR TITLE
Fix for issue 40774 RequestDelegateFactory null param default

### LINKED ISSUE #40774: Exception thrown in RequestDelegateFactory.Create when handling null as a default value for nullable struct type
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

An exception is thrown when supplying a null default for a parameter when using  RequestDelegateFactory.Create for a method like: `public void Myfunc(TimeSpan? estRunTime = null)`

The exception is thrown within:

private static Expression BindParameterFromValue(ParameterInfo parameter, Expression valueExpression, FactoryContext factoryContext, string source)

in this code
`
           var ifNotNullTryParse = !parameter.HasDefaultValue ?
                Expression.IfThen(TempSourceStringNotNullExpr, tryParseExpression) :
                Expression.IfThenElse(TempSourceStringNotNullExpr,
                    tryParseExpression,
                    Expression.Assign(argument, Expression.Constant(parameter.DefaultValue)));
`
The Assign throws because the argument type is incompatible with the type of the constant value.

Naively it appears to occur because in the Expression.Constant(parameter.DefaultValue) the parameter.DefaultValue is always declared as a  "object" type in  MethodInfo. In this scenario its a null value so Expression.Constant(...) can't determine the actual type from the value. Hence naively the solution would appear to be to use the version of Expression.Constant(...) which takes a second value specifying the type of the parameter. ie
`
                    Expression.Assign(argument, Expression.Constant(paramaer.ParameterType,parameter.DefaultValue)));
`
The implicit assumption here being that a MethodInfo parameter DefaultValue should always be of the the type ParameterType.

### Expected Behavior

The expectation is that this should be able to correctly determine and handle null as a valid default value for the function parameter.

### Steps To Reproduce

void Main()
{
	var mi = this.GetType().GetMethod("Myfunc");
	var rd = RequestDelegateFactory.Create(mi, hc => this);
}

public void Myfunc(TimeSpan? estRunTime = null) {}

### Exceptions (if any)
ArgumentException
Expression of type 'System.Object' cannot be used for assignment to type 'System.Nullable`1[System.TimeSpan]'

### Version
6.0.3

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Acerby**
# Fix for issue 40774 RequestDelegateFactory null param default

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description

Fix issue with default value for nullable struct parameters

Added missing parameter type info for expression tree constant

Also added test to verify this functionality

Fixes #40774  (in this specific format)

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 1]
```diff
@@ -387,7 +387,7 @@ public async Task RequestDelegatePopulatesFromNullableNullOptionalParameter()
 
         await requestDelegate(httpContext);
 
-        Assert.Equal(null, httpContext.Items["input"]);
+        Assert.IsNull(httpContext.Items["input"]);
```
Comment (campersau) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:390]:
```suggestion
        Assert.Null(httpContext.Items["input"]);
```

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 2]
```diff
@@ -226,6 +226,11 @@ private static void TestOptionalNullable(HttpContext httpContext, int? value = 4
         httpContext.Items.Add("input", value);
     }
 
+    private static void TestOptionalNullableNull(HttpContext httpContext, double? value = null)
+    {
+        httpContext.Items.Add("input", value);
```
Comment (halter73) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:231]:
```suggestion
        httpContext.Items.Add("input", value ?? "null");
```

The `ItemsDictionary` is unusual in that it returns `null` for missing values, so it's impossible to know from a test if this line executed or not.

https://github.com/dotnet/aspnetcore/blob/e001dc01a152806a457eb9da15dc102107451dee/src/Http/Http/src/Internal/ItemsDictionary.cs#L23-L33

Comment (Acerby) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:231]: Many thanks for spotting that @halter73. I have updated the Acerby:main branch based on a slightly updated version your suggestion of using "null".

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
@Acerby thanks for both finding and fixing this issue!

**Pull request merged**